feature/bgp/gracefulrestart/otg_tests/exrr/exrr_test.go

```diff
@@ -0,0 +1,1872 @@
+	flowipv6 := config.Flows().Add().SetName(srcip + "-" + dstip + "-Ipv6")
+	flowipv6.Metrics().SetEnable(true)
+	flowipv6.TxRx().Device().
+		SetTxNames([]string{src.Name + ".IPv6"}).
+		SetRxNames([]string{dst.Name + ".IPv6"})
+	flowipv6.Size().SetFixed(512)
+	flowipv6.Rate().SetPps(100)
+	e1 := flowipv6.Packet().Add().Ethernet()
+	e1.Src().SetValue(src.MAC)
+	v6 := flowipv6.Packet().Add().Ipv6()
+	v6.Src().SetValue(srcip)
+	v6.Dst().Increment().SetStart(dstip)
+}
+
+func verifyNoPacketLoss(t *testing.T, ate *ondatra.ATEDevice, flows []string) {
+	t.Helper()
+	otg := ate.OTG()
+	c := otg.FetchConfig(t)
+	otgutils.LogFlowMetrics(t, otg, c)
+
+	for _, f := range flows {
+		t.Logf("Verifying flow metrics for flow %s\n", f)
+		recvMetric := gnmi.Get(t, otg, gnmi.OTG().Flow(f).State())
+		txPackets := float32(recvMetric.GetCounters().GetOutPkts())
+		rxPackets := float32(recvMetric.GetCounters().GetInPkts())
+		lostPackets := txPackets - rxPackets
+		if txPackets == 0 {
+			t.Fatalf("Tx packets should be higher than 0 for flow %s", f)
+		}
+		if lossPct := lostPackets * 100 / txPackets; lossPct < 5.0 {
+			t.Logf("Traffic received as expected! Got %v loss", lossPct)
+		} else {
+			t.Errorf("traffic verification failed, Loss Pct for Flow %s: got %f", f, lossPct)
+		}
+	}
+}
+
+func confirmPacketLoss(t *testing.T, ate *ondatra.ATEDevice, flows []string) {
+	t.Helper()
+	otg := ate.OTG()
+	c := otg.FetchConfig(t)
+	otgutils.LogFlowMetrics(t, otg, c)
+	for _, f := range flows {
+		t.Logf("Verifying flow metrics for flow %s\n", f)
+		recvMetric := gnmi.Get(t, otg, gnmi.OTG().Flow(f).State())
+		txPackets := float32(recvMetric.GetCounters().GetOutPkts())
+		rxPackets := float32(recvMetric.GetCounters().GetInPkts())
+		lostPackets := txPackets - rxPackets
+		if txPackets == 0 {
+			t.Fatalf("Tx packets should be higher than 0 for flow %s", f)
+		}
+		if lossPct := lostPackets * 100 / txPackets; lossPct > 99.0 {
+			t.Logf("Traffic received as expected! Loss seen as expected: got %v, want 100%% ", lossPct)
+		} else {
+			t.Errorf("traffic %s is expected to fail: got %f, want 100%% failure", f, lossPct)
+		}
+	}
+}
+
+func sendTraffic(t *testing.T, ate *ondatra.ATEDevice) {
+	t.Helper()
+	t.Log("Starting traffic")
+	ate.OTG().StartTraffic(t)
+	time.Sleep(trafficDuration)
+	t.Log("Stop traffic")
+	ate.OTG().StopTraffic(t)
+}
+
+// createGracefulRestartAction create a bgp control action for initiating the graceful restart process
+func createGracefulRestartAction(t *testing.T, peerNames []string, restartDelay uint32, notification string) gosnappi.ControlAction {
+	t.Helper()
+	grAction := gosnappi.NewControlAction()
+	switch notification {
+	case "soft":
+		grAction.Protocol().Bgp().InitiateGracefulRestart().
+			SetPeerNames(peerNames).SetRestartDelay(restartDelay).Notification().Cease().SetSubcode(gosnappi.DeviceBgpCeaseErrorSubcode.ADMIN_RESET_CODE6_SUBCODE4)
+	case "hard":
```

Comment:
Following from RT-1.35.6 of the readme, where in the code, the following is checked on the ATE?


As per [rfc8538#section-3.1](https://datatracker.ietf.org/doc/html/rfc8538#section-3.1), when "N bit" exchanged between peers (i.e. GR negotiated), the "HARD RESET" notification of code 6 subcode 9 must be sent to the peer. However, the subcode for "Administrative Reset" i.e. code 6 subcode 4 must be carried in the data portion of subcode 9 notification message.
On receipt of the "HARD RESET" Notification message from the DUT, the ATEs MUST validate the Notification message and ensure it matches the bullet above